ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore/Lockstep.hs

```diff
@@ -0,0 +1,811 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Test.Ouroboros.Storage.LedgerDB.V1.BackingStore.Lockstep (
+    -- * Facilitate running the tests in @'IO'@ or @'IOSim'@.
+    IOLikeMonad (..)
+  , IOLikeMonadC (..)
+  , RealMonad
+  , unIOLikeMonad
+    -- * Model state
+  , BackingStoreState (..)
+  , RealEnv (..)
+  , maxOpenValueHandles
+  ) where
+
+import           Cardano.Slotting.Slot
+import           Control.Concurrent.Class.MonadMVar.Strict
+import           Control.Monad
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.IOSim
+import           Control.Monad.Reader
+import           Data.Bifunctor
+import           Data.Constraint
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Typeable
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore as BS
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore.Impl.InMemory as BS
+import           Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore.Impl.LMDB as LMDB
+                     (LMDBErr (..))
+import           Ouroboros.Consensus.Util.IOLike hiding (MonadMask (..),
+                     StrictMVar, handle, readMVar, swapMVar)
+import           System.FS.API hiding (Handle)
+import qualified System.FS.API.Types as FS
+import           Test.Cardano.Ledger.Binary.Arbitrary ()
+import qualified Test.Ouroboros.Storage.LedgerDB.V1.BackingStore.Mock as Mock
+import           Test.Ouroboros.Storage.LedgerDB.V1.BackingStore.Mock (Err (..),
+                     Mock (..), ValueHandle (..), runMockState)
+import           Test.Ouroboros.Storage.LedgerDB.V1.BackingStore.Registry
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck (Gen)
+import           Test.QuickCheck.StateModel
```

Comment:
TODO I couldn't review this particularly well, since I'm unfamilar with `Test.QuickCheck.StateModel`.

Reply:
Note that these tests use `quickcheck-lockstep`, so if you want to understand the framework I'd start reading the documentation there instead of `quickcheck-dynamic`